cmd/result-aggregator/passwd.go

```diff
@@ -0,0 +1,54 @@
+package main
+
+import (
+	"encoding/csv"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+)
+
+type passwdFile struct {
+	username      string
+	password      func() []byte
```

Comment:
This is not part of the file. Just have two separate validators that implement your interface. One uses a literal password, one reads the htpasswd file. Combine them with logical OR.

Reply:
Similar (but not the same) problem is solved [here](https://github.com/openshift/ci-tools/blob/aa5a6f46442730c79c61708c61b60d2c7b6f4d55/pkg/slack/events/handler.go#L80-L93)

Reply:
So I'd expect to see:

```go

type validator interface {
	Validate(username, password string) bool
}

type htpasswd struct {
	file string
}

func (h *htpasswd) Validate(username, password string) bool {
	reader, err := os.Open(h.file)
	if err != nil {
		logrus.WithField("file", h.file).Error("failed to open file")
		return false
	}
	defer func() {
		if err := reader.Close(); err != nil {
			logrus.WithError(err).Warn("Could not close htpasswd file reader.")
		}
	}()

	csvReader := csv.NewReader(reader)
	csvReader.Comma = ':'
	csvReader.Comment = '#'
	csvReader.TrimLeadingSpace = true

	records, err := csvReader.ReadAll()
	if err != nil {
		logrus.Error("failed to read csv")
	}

	for _, record := range records {
		if len(record) == 2 && record[0] == username && record[1] == password {
			return true
		}
	}
	return false
}

type literal struct {
	username, password string
}

func (l *literal) Validate(username, password string) bool {
	return username == l.username && password == l.password
}

type multi struct {
	delegates []validator
}

func (m *multi) Validate(username, password string) bool {
	for _, delegate := range m.delegates {
		if delegate.Validate(username, password) {
			return true
		}
	}
	return false
}
```

NOTE: handle errors for `Close()` in your `defer`. Change your editor to catch this.

Reply:
I see. Updated.

> NOTE: handle errors for Close() in your defer. Change your editor to catch this.

It is enabled by default in GoLand. And my eyes did not catch that out.